pkg/syncer/storagepool/intended_state.go

```diff
@@ -181,23 +183,36 @@ func (c *SpController) applyIntendedState(ctx context.Context, state *intendedSt
 		if ok && statusErr.Status().Reason == metav1.StatusReasonNotFound {
 			log.Infof("Creating StoragePool instance for %s", state.spName)
 			sp := state.createUnstructuredStoragePool(ctx)
+			// Generation Number of a newly created object is 1. After issuing a write operation the
+			// current thread typically sleeps and wait for server reply. Meanwhile SPWatch can receive
+			// the write event and start to detect if this event was due to an external change. So it is
+			// essential to store expected value in SPNameToGenerationNumberMap before issuing any write
+			// operation to StoragePool to avoid false positive.
+			c.spWatcher.SPNameToGenerationNumberMap[state.spName] = 1
 			newSp, err := spClient.Resource(*spResource).Create(ctx, sp, metav1.CreateOptions{})
 			if err != nil {
+				delete(c.spWatcher.SPNameToGenerationNumberMap, state.spName)
 				log.Errorf("Error creating StoragePool %s. Err: %+v", state.spName, err)
 				return err
 			}
+			c.spWatcher.SPNameToGenerationNumberMap[newSp.GetName()] = newSp.GetGeneration()
 			log.Debugf("Successfully created StoragePool %v", newSp)
 		}
 	} else {
 		// StoragePool already exists, so Update it
 		// We don't expect ConflictErrors since updates are synchronized with a lock
 		log.Infof("Updating StoragePool instance for %s", state.spName)
 		sp := state.updateUnstructuredStoragePool(ctx, sp)
+		// On each modification generation number increases by 1.
+		c.spWatcher.SPNameToGenerationNumberMap[state.spName]++
```

Comment:
In case of syncer restarts, isn't this map going to be empty? This will lead to a null pointer exception... which will cause the syncer to restart again.

Reply:
Yes this map can be empty **but not nil** at this moment. Before starting SpController we first initialize SPNameToGenerationNumberMap to avoid this null pointer exception. In golang accessing key which does not exist in a non nil map does not results in panic but returns default value (0 for integers). I have insured that before any access to the map it is initialized.
```   go
// Start the services
spWatcher := newStoragePoolWatch()

spController, err := newSPController(vc, configInfo.Cfg.Global.ClusterID, spWatcher)
```

```go
// newStoragePoolWatch initialises a SPWatcher instance.
func newStoragePoolWatch() *SPWatcher {
	w := &SPWatcher{}
	w.SPNameToGenerationNumberMap = make(map[string]int64)
	return w
}
```